builtin/common/vector.lua

```diff
@@ -141,3 +141,82 @@ function vector.sort(a, b)
 	return {x = math.min(a.x, b.x), y = math.min(a.y, b.y), z = math.min(a.z, b.z)},
 		{x = math.max(a.x, b.x), y = math.max(a.y, b.y), z = math.max(a.z, b.z)}
 end
+
+local function sin(x)
+	if x % math.pi == 0 then
+		return 0
+	else
+		return math.sin(x)
+	end
+end
+
+local function cos(x)
+	if x % math.pi == math.pi / 2 then
+		return 0
+	else
+		return math.cos(x)
+	end
+end
+
+function vector.rotate_around_axis(v, axis, angle)
+	local cosangle = cos(angle)
+	local sinangle = sin(angle)
+	axis = vector.normalize(axis)
+	-- https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula
+	local dot_axis = vector.multiply(axis, vector.dot(axis, v))
+	local cross = vector.cross(axis, v)
+	return vector.new(
+		cross.x * sinangle + (v.x - dot_axis.x) * cosangle + dot_axis.x,
+		cross.y * sinangle + (v.y - dot_axis.y) * cosangle + dot_axis.y,
+		cross.z * sinangle + (v.z - dot_axis.z) * cosangle + dot_axis.z
+	)
+end
+
+function vector.rotate(v, rot)
+	local sinpitch = sin(-rot.x)
+	local sinyaw = sin(-rot.y)
+	local sinroll = sin(-rot.z)
+	local cospitch = cos(rot.x)
+	local cosyaw = cos(rot.y)
+	local cosroll = math.cos(rot.z)
+	-- rotation matrix that applies yaw, pitch and roll
+	-- result of matrix multiplying the rotation matrices for yaw, pitch and roll
+	local matrix = {
+		{
+			cosyaw * cospitch,
+			cosyaw * sinpitch * sinroll - sinyaw * cosroll,
+			cosyaw * sinpitch * cosroll + sinyaw * sinroll
+		},
+		{
+			sinyaw * cospitch,
+			sinyaw * sinpitch * sinroll + cosyaw * cosroll,
+			sinyaw * sinpitch * cosroll - cosyaw * sinroll
+		},
+		{
+			-sinpitch,
+			cospitch * sinroll,
+			cospitch * cosroll
+		},
+	}
+	-- compute matrix multiplication: `matrix` * `v`
+	return vector.new(
+		matrix[2][2] * v.x + matrix[2][3] * v.y + matrix[2][1] * v.z,
+		matrix[3][2] * v.x + matrix[3][3] * v.y + matrix[3][1] * v.z,
+		matrix[1][2] * v.x + matrix[1][3] * v.y + matrix[1][1] * v.z
+	)
+end
+
+function vector.directions_to_rotation(forward, up)
+	local rot = {x = math.asin(forward.y), y = -math.atan2(forward.x, forward.z), z = 0}
+	-- calculate vector pointing uwith roll = 0, just based on forward vector
+	local forwup = vector.rotate({x = 0, y = 1, z = 0}, rot)
+	-- 'forwup' and 'up' are now in a plane with 'forward' as normal.
+	-- the angle between them is equal to math.abs(roll)
+	rot.z = vector.angle(forwup, up)
+	-- if 'up' rotated by rot.z is equal to forwup, then roll = -rot.z, else roll = rot.z
```

Comment:
This comment states the obvious, more helpful would be a short explanation <i>why</i> rot.z needs to be inverted in some cases